contracts/cw-fund-distributor/src/contract.rs

```diff
@@ -0,0 +1,621 @@
+        .add_attribute("total_power", total_power.power))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Receive(cw20::Cw20ReceiveMsg {
+            sender: _,
+            amount,
+            msg: _,
+        }) => execute_receive_cw20(deps, info.sender, amount),
+        ExecuteMsg::Fund {} => execute_fund_natives(deps, info),
+        ExecuteMsg::ClaimCw20s { tokens } => execute_claim_cw20s(deps, info.sender, tokens),
+        ExecuteMsg::ClaimNatives { denoms } => execute_claim_natives(deps, info.sender, denoms),
+        ExecuteMsg::WithdrawCw20s { tokens } => execute_withdraw_cw20s(deps, info.sender, tokens),
+        ExecuteMsg::WithdrawNatives { denoms } => {
+            execute_withdraw_natives(deps, info.sender, denoms)
+        }
+        ExecuteMsg::UpdateAdmin { admin } => execute_update_admin(deps, info.sender, admin),
+    }
+}
+
+/// Computes the total entitlement for the provided set of
+/// paramater. NOTE: this may differ than the actual amount that an
+/// address is entitled to if the address has already claimed some of
+/// their entitlement.
+fn compute_entitled(
+    provided: Uint128,
+    voting_power: Uint128,
+    total_power: Uint128,
+) -> StdResult<Uint128> {
+    Ok(provided
+        .full_mul(voting_power)
+        .checked_div(Uint256::from(total_power))?
+        .try_into()
+        .unwrap())
+}
+
+/// Generic function for computing the amount of tokens an address is
+/// entitled to for each denom being distributed.
+fn compute_entitlements<'a, K>(
+    deps: Deps,
+    contract_balances: Vec<(K, Uint128)>,
+    sender: Addr,
+    claims: Map<'a, (Addr, K), Uint128>,
+) -> StdResult<Vec<(K, Uint128)>>
+where
+    (Addr, K): PrimaryKey<'a>,
+    K: Clone,
+{
+    let total_power = TOTAL_POWER.load(deps.storage)?;
+    let dist_height = DISTRIBUTION_HEIGHT.load(deps.storage)?;
+    let voting_contract = VOTING_CONTRACT.load(deps.storage)?;
+    let voting_power: voting::VotingPowerAtHeightResponse = deps.querier.query_wasm_smart(
+        voting_contract,
+        &voting::Query::VotingPowerAtHeight {
+            address: sender.to_string(),
+            height: Some(dist_height),
+        },
+    )?;
+
+    contract_balances
+        .into_iter()
+        .map(|(key, provided)| {
+            let total_entitlement = compute_entitled(provided, voting_power.power, total_power)?;
+            let claimed = claims
+                .may_load(deps.storage, (sender.clone(), key.clone()))?
+                .unwrap_or_default();
+            let entitled = total_entitlement - claimed;
+            Ok((key, entitled))
+        })
+        .collect::<StdResult<Vec<_>>>()
+}
+
+/// Generic function for paginating a list of (K, V) pairs in a
+/// CosmWasm Map.
+fn paginate_items<'a, K, V>(
```

Comment:
i can totally see us wanting to move this to a shared utils at a later date

Reply:
yes!